lib/revision.rb

```diff
@@ -0,0 +1,106 @@
+class Metis
+  class Revision
+    FILENAME_MATCH=/[^<>:;,?"*\|\/\x00-\x1f]+/x
+
+    FILEPATH_MATCH=%r!
+      \A
+        metis:\/\/
+        (?<project_name>(#{FILENAME_MATCH.source}))
+        /
+        (?<bucket_name>(#{FILENAME_MATCH.source}))
+        /
+        (?<file_path>(#{FILENAME_MATCH.source}).*)
+      \z
+    !x
+
+    attr_reader :source, :dest
+    def initialize(params)
+      raise Etna::BadRequest, 'All revisions require "source" parameter' unless params.has_key? :source
+      raise Etna::BadRequest, 'All revisions require "dest" parameter' unless params.has_key? :dest
+
+      raise Etna::BadRequest, 'Invalid source path' unless valid_file_path?(params[:source])
+
+      @source = params[:source]
+      @dest = params[:dest]
+      # TODO: Will have to handle DELETE differently,
+      #       when params[:dest] will be nil
+    end
+
+    def self.create_from_parts(params)
+      Metis::Revision.new({
+        source: path_from_parts(
+          params[:source][:project_name],
+          params[:source][:bucket_name],
+          params[:source][:file_path]),
+        dest: path_from_parts(
+          params[:dest][:project_name],
+          params[:dest][:bucket_name],
+          params[:dest][:file_path])
+      })
+    end
+
+    def self.extract_bucket_name_from_path(path)
+      # Assumes Metis file paths are in the form
```

Comment:
Not to go too deep down the rabbit hole of abstraction here, but since we are committing to this "metis path" notion, it probably deserves its own service object, e.g. `MetisPath.new(path).bucket_name`

Reply:
A lot of the methods before could go away if we had `@source = MetisPath.new(source_path)`.